src/Psy/Readline/TabCompletion.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Psy\Readline;
+
+use Psy\Context;
+
+/**
+ * Class TabCompletion
+ * @package Psy\Readline
+ */
+class TabCompletion
+{
+    /** @var Context */
+    protected $context;
+
+    public function __construct(Context $context = null)
+    {
+        $this->context = $context;
+    }
+
+    public function activate()
+    {
+        readline_completion_function(array(&$this, 'callback'));
+    }
+
+    /**
+     * @param $input
+     * @param $index
+     * @return array
+     */
+    public function callback($input, $index)
+    {
+        $info = readline_info();
+
+        // the char just before the cursor is a dollar? send a variable context
+        $charAt = substr($info['line_buffer'], $index-1, 1);
+        if ($charAt == '$') {
+            return array_keys($this->context->getAll());
+        }
+
+        $parenthesize = function ($name) {
+            return sprintf('%s()', $name);
+        };
+
+        if (strlen($info['line_buffer']) > 4) {
```

Comment:
potential for bugs here: `$info['line_buffer']` seems to be quite literally the buffer they use for input, and it often contains things that aren't part of the current line (e.g. left over from the last line). We should do `$line = substr($info['line_buffer'], 0, $info['end'])` somewhere early on in this function and always refer to that instead.